### PR TITLE
Correct some bugs in training / evaluating final model

### PR DESCRIPTION
**Pull request by ebianchi**
This PR combines fixes to a couple of minor bugs, plus adds a couple of minor features.  All of these are related to the training process, resuming experiments, and generating results from a completed experiment.

### Bug 1:  Evaluating a system that cannot handle batched data
An assertion statement in `system.py` ([line here](https://github.com/DAIRLab/dair_pll/blob/main/dair_pll/system.py#L115)) intended to check if there was a maximum batch dimension, then to calculate if the provided data exceeded that maximum.  As written, a maximum batch dimension of zero is treated the same as if an unset maximum batch dimension.  The latter should not impose a limit, whereas the former should.

**Fix:**  A quick change to the one line of code:  `if self.max_batch_dim` --> `if self.max_batch_dim is not None`.

### Bug 2:  Training an already finished experiment
When training an experiment which has already finished training (whether by early stopping or by reaching an epoch maximum), `experiment.py` throws an error when it attempts to terminate `train()` because the return argument `training_loss` is not set.

**Fix:** The `training_loss` variable is set before attempting further training via a call to a new `calculate_loss_no_grad_step()` method of the `Experiment` class.  This new method is the same as `train_epoch()` but without any gradient steps, and it is called after setting the model in `eval` mode.

### New feature 1:  Record if training has completed or not
An experiment's `TrainingState` now has a `finished_training` flag, which is `False` until the training process stops due to hitting the maximum epochs or from early stopping.

### New feature 2:  Log model performance with initialized parameters
Before, the model's initialized parameters were never a candidate for the best learned system state.  Now, training an experiment starts with evaluating the model's training and validation losses without taking any gradient steps, initializing the stored best validation loss with this evaluation.

### CODE REVIEW

File: dair_pll/experiment.py [hunk 1]
```diff
@@ -340,6 +342,26 @@ def train_epoch(self, data: DataLoader, system: System,
         avg_loss = cast(Tensor, sum(losses) / len(losses))
         return avg_loss
 
+    def calculate_loss_no_grad_step(self, data: DataLoader, system: System) \
```
Comment (mshalm) [inline dair_pll/experiment.py]: Seems almost copy-paste from `train_epoch`. Probably better to have 1 function that takes a flag as an argument for whether to take the gradient steps or not -- e.g. by making the `Optimizer` argument to `train_epoch` an `Optional[Optimizer]`, and take gradient steps iff it's not `None`.

File: examples/contactnets_simple.py [hunk 2]
```diff
@@ -225,9 +225,15 @@ def regenerate_callback(epoch: int, learned_system: System,
         cast(MultibodyLearnableSystem, learned_system).generate_updated_urdfs()
 
     # Trains system.
+    print("Training.")
     experiment.train(
         regenerate_callback if regenerate else default_epoch_callback)
 
+    # Store the results of the trained system.
```
Comment (mshalm) [inline examples/contactnets_simple.py]: The intended usage of `get_results` is that you don't have to call `train` first -- it calls `train` for you.

Up for debate if the method is properly named given that. Maybe should be something more like `do_experiment`, `calculate_results`, `train_and_evaluate`, `run`, etc

**Review by mshalm — CHANGES_REQUESTED** [review]
Overall clean fix of two significant bugs. Two minor comments.

File: examples/contactnets_simple.py (hunk 2 above)
Comment (ebianchi) [reply examples/contactnets_simple.py]: I see.  That's not clear since they don't return the same things.  E.g. if you wanted to get the statistics and the last training loss, you'd have do call `train` then `get_results` anyway.

File: dair_pll/experiment.py (hunk 1 above)
Comment (ebianchi) [reply dair_pll/experiment.py]: Agreed.  Done just like your suggestion.

File: examples/contactnets_simple.py (hunk 2 above)
Comment (ebianchi) [reply examples/contactnets_simple.py]: After some discussion, the workable solution we came to is the following:

- `get_results` is renamed `generate_results` to indicate more clearly that it will start/continue the training process as necessary.
- The newly named `generate_results` now returns both the learned model corresponding to the best validation loss and the final evaluation statistics.
- The example script `contactnets_simple.py` calls `generate_results` only, not also `train`.

These changes are implemented in the latest commit in this PR.

File: dair_pll/experiment.py [hunk 3]
```diff
@@ -781,23 +812,29 @@ def _evaluation(self, learned_system: System) -> StatisticsDict:
...
 
-    def get_results(
+    def generate_results(
         self,
         epoch_callback: EpochCallbackCallable = default_epoch_callback,
-    ) -> StatisticsDict:
-        r"""Gets final results/statistics of experiment. This will return
-        previously saved results on disk if they already exist, or run the
-        experiment to generate them if they don't.
+    ) -> Tuple[Tensor, StatisticsDict]:
```
Comment (mshalm) [inline dair_pll/experiment.py]: Should be Tuple[System, StatisticsDict]:

**Review by mshalm — APPROVED** [review]
LGTM -- minus one incorrect type hint.

**Review by mshalm — APPROVED** [review]
LGTM -- minus one incorrect type hint.

**Comment by ebianchi** [comment]
> LGTM -- minus one incorrect type hint.

Corrected!  Thanks for the catch.
